### ISSUE TITLE
It seems that this SFT code is not performed only on the answer, but on the entire sequence?

### ISSUE THREAD

**Issue by deeplearning-LI** [huggingface/open-r1#652]
I didn't find any label-related processing in SFT.

`

    def _prepare_dataset(
        self,
        dataset: Union[Dataset, IterableDataset],
        processing_class: Union[PreTrainedTokenizerBase, BaseImageProcessor, FeatureExtractionMixin, ProcessorMixin],
        args: SFTConfig,
        packing: bool,
        formatting_func: Optional[Callable[[dict], str]],
        dataset_name: str,) 

        if isinstance(dataset, ConstantLengthDataset):
            return dataset

        # If the dataset is already preprocessed (tokenized), skip the processing steps.
        column_names = list(next(iter(dataset)).keys())
        is_processed = "input_ids" in column_names

        # Build the kwargs for the `map` function
        map_kwargs = {}
        if isinstance(dataset, Dataset):  # IterableDataset does not support num_proc
            map_kwargs["num_proc"] = args.dataset_num_proc

        with PartialState().local_main_process_first():
            # Apply the formatting function if any
            if formatting_func is not None and is_processed:
                warnings.warn(
                    "You passed a dataset that is already processed (contains an `input_ids` field) together with a "
                    "formatting function. Therefore `formatting_func` will be ignored. Either remove the "
                    "`formatting_func` or pass a dataset that is not already processed.",
                    UserWarning,
                )

            if formatting_func is not None and not is_processed:
                if isinstance(dataset, Dataset):  # `IterableDataset.map` does not support `desc`
                    map_kwargs["desc"] = f"Applying formatting function to {dataset_name} dataset"

                batched = isinstance(formatting_func(next(iter(dataset))), list)

                def _func(example):
                    return {"text": formatting_func(example)}

                dataset = dataset.map(_func, batched=batched, **map_kwargs)

            # If the dataset is prompt-completion, convert it to language modeling type
            # The default is messages, so that they can be processed uniformly later.
            if "prompt" in dataset.column_names and "completion" in dataset.column_names:
                key = "messages" if is_conversational(dataset[0]) else "text"

                def concat_prompt_completion(example):
                    return {key: example["prompt"] + example["completion"]}

                dataset = dataset.map(concat_prompt_completion, remove_columns=["prompt", "completion"])

            # Convert the dataset to ChatML if needed
            if isinstance(dataset, Dataset):  # `IterableDataset.map` does not support `desc`
                map_kwargs["desc"] = f"Converting {dataset_name} dataset to ChatML"
            dataset = dataset.map(
                maybe_convert_to_chatml,
                remove_columns="conversations" if "conversations" in dataset.column_names else None,
                **map_kwargs,
            )

            # Apply the chat template if needed
            if isinstance(dataset, Dataset):  # `IterableDataset.map` does not support `desc`
                map_kwargs["desc"] = f"Applying chat template to {dataset_name} dataset"
            dataset = dataset.map(
                maybe_apply_chat_template,
                fn_kwargs={"tokenizer": processing_class},
                remove_columns="messages" if "messages" in dataset.column_names else None,
                **map_kwargs,
            )

            # Tokenize the dataset if needed
            if not is_processed:
                if isinstance(dataset, Dataset):  # `IterableDataset.map` does not support `desc`
                    map_kwargs["desc"] = f"Tokenizing {dataset_name} dataset"

                def tokenize(example, processing_class, dataset_text_field):
                    return processing_class(example[dataset_text_field])

                dataset = dataset.map(
                    tokenize,
                    fn_kwargs={"processing_class": processing_class, "dataset_text_field": args.dataset_text_field},
                    **map_kwargs,
                )

            # Pack or truncate
            if packing:
                if args.max_seq_length is None:
                    raise ValueError("When packing is enabled, `max_seq_length` can't be `None`.")
                if isinstance(dataset, Dataset):  # `IterableDataset.map` does not support `desc`
                    map_kwargs["desc"] = f"Packing {dataset_name} dataset"
                dataset = dataset.select_columns("input_ids")
                dataset = dataset.map(
                    pack_examples, batched=True, fn_kwargs={"seq_length": args.max_seq_length}, **map_kwargs
                )
            elif args.max_seq_length is not None:
                if isinstance(dataset, Dataset):  # `IterableDataset.map` does not support `desc`
                    map_kwargs["desc"] = f"Truncating {dataset_name} dataset"

                def truncate(example, max_seq_length):
                    return {key: example[key][:max_seq_length] for key in ["input_ids", "attention_mask"]}

                dataset = dataset.map(
                    truncate,
                    fn_kwargs={"max_seq_length": args.max_seq_length},
                    **map_kwargs,
                )

            # For Liger kernel, ensure only input_ids is present
            if args.use_liger:
                dataset = dataset.select_columns("input_ids")

        return dataset

    def compute_loss(self, model, inputs, return_outputs=False, num_items_in_batch=None):
        """
        Compute training loss and additionally compute token accuracies
        """
        (loss, outputs) = super().compute_loss(
            model, inputs, return_outputs=True, num_items_in_batch=num_items_in_batch
        )

        # Compute token accuracy if we have labels and if the model is not using Liger (no logits)
        if "labels" in inputs and not self.args.use_liger:
            shift_logits = outputs.logits[..., :-1, :].contiguous()
            shift_labels = inputs["labels"][..., 1:].contiguous()

            # Get predictions
            predictions = shift_logits.argmax(dim=-1)

            # Create mask for non-padding tokens (assuming ignore_index is -100)
            mask = shift_labels != -100

            # Calculate accuracy only on non-padding tokens
            correct_predictions = (predictions == shift_labels) & mask
            total_tokens = mask.sum()
            correct_tokens = correct_predictions.sum()

            # Gather the correct_tokens and total_tokens across all processes
            correct_tokens = self.accelerator.gather_for_metrics(correct_tokens)
            total_tokens = self.accelerator.gather_for_metrics(total_tokens)

            # Compute the mean token accuracy and log it
            accuracy = (correct_tokens.sum() / total_tokens.sum()).item() if total_tokens.sum() > 0 else 0.0
            self._metrics["mean_token_accuracy"].append(accuracy)

        return (loss, outputs) if return_outputs else loss

`

**How to add label to this code?**

**Comment by deeplearning-LI** [comment]
@lewtun

**Comment by deeplearning-LI** [comment]
> I didn't find any label-related processing in SFT.
> 
> `
> 
> ```
> def _prepare_dataset(
>     self,
>     dataset: Union[Dataset, IterableDataset],
>     processing_class: Union[PreTrainedTokenizerBase, BaseImageProcessor, FeatureExtractionMixin, ProcessorMixin],
>     args: SFTConfig,
>     packing: bool,
>     formatting_func: Optional[Callable[[dict], str]],
>     dataset_name: str,) 
> 
> 
> 
> 
>     if isinstance(dataset, ConstantLengthDataset):
>         return dataset
> 
> 
> 
>     # If the dataset is already preprocessed (tokenized), skip the processing steps.
>     column_names = list(next(iter(dataset)).keys())
>     is_processed = "input_ids" in column_names
> 
>     # Build the kwargs for the `map` function
>     map_kwargs = {}
>     if isinstance(dataset, Dataset):  # IterableDataset does not support num_proc
>         map_kwargs["num_proc"] = args.dataset_num_proc
> 
>     with PartialState().local_main_process_first():
>         # Apply the formatting function if any
>         if formatting_func is not None and is_processed:
>             warnings.warn(
>                 "You passed a dataset that is already processed (contains an `input_ids` field) together with a "
>                 "formatting function. Therefore `formatting_func` will be ignored. Either remove the "
>                 "`formatting_func` or pass a dataset that is not already processed.",
>                 UserWarning,
>             )
> 
>         if formatting_func is not None and not is_processed:
>             if isinstance(dataset, Dataset):  # `IterableDataset.map` does not support `desc`
>                 map_kwargs["desc"] = f"Applying formatting function to {dataset_name} dataset"
> 
>             batched = isinstance(formatting_func(next(iter(dataset))), list)
> 
>             def _func(example):
>                 return {"text": formatting_func(example)}
> 
>             dataset = dataset.map(_func, batched=batched, **map_kwargs)
> 
>         # If the dataset is prompt-completion, convert it to language modeling type
>         # The default is messages, so that they can be processed uniformly later.
>         if "prompt" in dataset.column_names and "completion" in dataset.column_names:
>             key = "messages" if is_conversational(dataset[0]) else "text"
> 
>             def concat_prompt_completion(example):
>                 return {key: example["prompt"] + example["completion"]}
> 
>             dataset = dataset.map(concat_prompt_completion, remove_columns=["prompt", "completion"])
> 
>         # Convert the dataset to ChatML if needed
>         if isinstance(dataset, Dataset):  # `IterableDataset.map` does not support `desc`
>             map_kwargs["desc"] = f"Converting {dataset_name} dataset to ChatML"
>         dataset = dataset.map(
>             maybe_convert_to_chatml,
>             remove_columns="conversations" if "conversations" in dataset.column_names else None,
>             **map_kwargs,
>         )
> 
>         # Apply the chat template if needed
>         if isinstance(dataset, Dataset):  # `IterableDataset.map` does not support `desc`
>             map_kwargs["desc"] = f"Applying chat template to {dataset_name} dataset"
>         dataset = dataset.map(
>             maybe_apply_chat_template,
>             fn_kwargs={"tokenizer": processing_class},
>             remove_columns="messages" if "messages" in dataset.column_names else None,
>             **map_kwargs,
>         )
> 
>         # Tokenize the dataset if needed
>         if not is_processed:
>             if isinstance(dataset, Dataset):  # `IterableDataset.map` does not support `desc`
>                 map_kwargs["desc"] = f"Tokenizing {dataset_name} dataset"
> 
>             def tokenize(example, processing_class, dataset_text_field):
>                 return processing_class(example[dataset_text_field])
> 
>             dataset = dataset.map(
>                 tokenize,
>                 fn_kwargs={"processing_class": processing_class, "dataset_text_field": args.dataset_text_field},
>                 **map_kwargs,
>             )
> 
>         # Pack or truncate
>         if packing:
>             if args.max_seq_length is None:
>                 raise ValueError("When packing is enabled, `max_seq_length` can't be `None`.")
>             if isinstance(dataset, Dataset):  # `IterableDataset.map` does not support `desc`
>                 map_kwargs["desc"] = f"Packing {dataset_name} dataset"
>             dataset = dataset.select_columns("input_ids")
>             dataset = dataset.map(
>                 pack_examples, batched=True, fn_kwargs={"seq_length": args.max_seq_length}, **map_kwargs
>             )
>         elif args.max_seq_length is not None:
>             if isinstance(dataset, Dataset):  # `IterableDataset.map` does not support `desc`
>                 map_kwargs["desc"] = f"Truncating {dataset_name} dataset"
> 
>             def truncate(example, max_seq_length):
>                 return {key: example[key][:max_seq_length] for key in ["input_ids", "attention_mask"]}
> 
>             dataset = dataset.map(
>                 truncate,
>                 fn_kwargs={"max_seq_length": args.max_seq_length},
>                 **map_kwargs,
>             )
> 
>         # For Liger kernel, ensure only input_ids is present
>         if args.use_liger:
>             dataset = dataset.select_columns("input_ids")
> 
>     return dataset
> 
> def compute_loss(self, model, inputs, return_outputs=False, num_items_in_batch=None):
>     """
>     Compute training loss and additionally compute token accuracies
>     """
>     (loss, outputs) = super().compute_loss(
>         model, inputs, return_outputs=True, num_items_in_batch=num_items_in_batch
>     )
> 
>     # Compute token accuracy if we have labels and if the model is not using Liger (no logits)
>     if "labels" in inputs and not self.args.use_liger:
>         shift_logits = outputs.logits[..., :-1, :].contiguous()
>         shift_labels = inputs["labels"][..., 1:].contiguous()
> 
>         # Get predictions
>         predictions = shift_logits.argmax(dim=-1)
> 
>         # Create mask for non-padding tokens (assuming ignore_index is -100)
>         mask = shift_labels != -100
> 
>         # Calculate accuracy only on non-padding tokens
>         correct_predictions = (predictions == shift_labels) & mask
>         total_tokens = mask.sum()
>         correct_tokens = correct_predictions.sum()
> 
>         # Gather the correct_tokens and total_tokens across all processes
>         correct_tokens = self.accelerator.gather_for_metrics(correct_tokens)
>         total_tokens = self.accelerator.gather_for_metrics(total_tokens)
> 
>         # Compute the mean token accuracy and log it
>         accuracy = (correct_tokens.sum() / total_tokens.sum()).item() if total_tokens.sum() > 0 else 0.0
>         self._metrics["mean_token_accuracy"].append(accuracy)
> 
>     return (loss, outputs) if return_outputs else loss
> ```
> 
> `
> 
> **How to add label to this code?**

I saw there is a parameter to control that, and that solved my problem.

trl.SFTConfig --> completion_only_loss
